test/auto_parallel/pir/test_reshard.py

```diff
@@ -0,0 +1,101 @@
+
+import paddle
+import paddle.distributed as dist
+from paddle import nn
+
+sys.path.append(str(pathlib.Path(__file__).resolve().parents[0]))
+from test_to_static_pir_program import (
+    BATCH_SIZE,
+    IMAGE_SIZE,
+    DemoNet,
+    create_data_loader,
+)
+
+
+class ReshardDemoNet(DemoNet):
+    def __init__(self, mesh, shard=True):
+        super().__init__(mesh, shard=True)
+
+    def forward(self, x):
+        out = DemoNet.forward(self, x)
+        out = dist.reshard(out, self._mesh, [dist.Shard(0)])
+        return out
+
+
+class TestToStaticPirProgramTrain(unittest.TestCase):
+    def test_to_static_program(self):
+        mesh = dist.ProcessMesh([0, 1], dim_names=["x"])
+        layer = ReshardDemoNet(mesh)
+        opt = paddle.optimizer.SGD(
+            learning_rate=0.1, parameters=layer.parameters()
+        )
+        loss_fn = nn.MSELoss()
+        loader = create_data_loader()
+        dist_loader = dist.shard_dataloader(loader, meshes=[mesh])
+        dist_model = dist.to_static(layer, dist_loader, loss_fn, opt)
+        engine = dist_model._engine
+        engine._build("train")
+        dist_program = engine._fwd_main_progs["train"]
+        dist_program = paddle.base.libpaddle.pir.apply_mix2dist_pass(
+            dist_program
+        )
+        loss = dist_program.get_output_value_by_name(engine._loss_names[0])
+        with paddle.static.program_guard(dist_program):
+            params_grads = paddle.autograd.ir_backward.append_backward(loss)
+            engine._optimizer._apply_optimize(
+                loss, startup_program=None, params_grads=params_grads
+            )
+
+        index = 0
+        for op in dist_program.global_block().ops:
+            if op.name() == 'dist_op.reshard':
+                if index == 0:
+                    # forward reshard op
+                    self.fwd_input = op.operand_source(0)
+                    self.assertEqual(
+                        self.fwd_input.dist_attr().dims_mapping, [-1, -1]
+                    )
+                    self.assertEqual(
+                        self.fwd_input.dist_attr().partial_dims, set()
+                    )
+                    self.assertEqual(
+                        self.fwd_input._local_shape,
+                        [BATCH_SIZE, IMAGE_SIZE // 2],
+                    )
+                    self.fwd_output = op.result(0)
+                    self.assertEqual(
+                        self.fwd_output.dist_attr().dims_mapping, [0, -1]
+                    )
+                    self.assertEqual(
+                        self.fwd_output.dist_attr().partial_dims, set()
+                    )
+                    self.assertEqual(
+                        self.fwd_output._local_shape,
+                        [BATCH_SIZE / 2, IMAGE_SIZE // 2],
+                    )
+                elif index == 1:
+                    # backward reshard op
+                    self.assertEqual(op.result(0).type(), self.fwd_input.type())
```

Comment:
should also check the input of backward reshard 
type could not be equal, dist attr should be equal

Reply:
ok